go/libraries/doltcore/sqle/autoincr/auto_increment_tracker.go

```diff
@@ -0,0 +1,100 @@
+package autoincr
+
+import (
+	"fmt"
+	"sync"
+)
+
+type AutoIncrementTracker interface {
+	Request(tableName string, val interface{}) (bool, error)
+	Confirm(tableName string, val interface{})
+}
+
+// AutoIncrementTracker is a global map that tracks the next auto increment value for each table in each database.
+// It is primarily used for concurrent transactions on an auto incremented table.
+func NewAutoIncrementTracker() AutoIncrementTracker {
+	return &autoIncrementTracker{
+		tables:  make(map[string]interface{}),
+		written: make(map[string]bool),
+	}
+}
+
+type autoIncrementTracker struct {
+	tables  map[string]interface{}
+	written map[string]bool
+	mu      sync.Mutex
+}
+
+var _ AutoIncrementTracker = (*autoIncrementTracker)(nil)
+
+// Request is used by an integrator to request whether the wanted auto increment key is valid for use. Requests do not
+// necessarily update the tracker as an integrator may query the same auto increment value multiple times (in the
+// information schema for example). A request updates the tracker when it is greater than the current value stored for
+// the table.
+func (a *autoIncrementTracker) Request(tableName string, val interface{}) (bool, error) {
+	a.mu.Lock()
+	defer a.mu.Unlock()
+
+	newVal, err := ConvertIntTypeToUint(val)
+	if err != nil {
+		return false, err
+	}
+
+	currVal := a.tables[tableName]
+	written := a.written[tableName]
+
+	if !written {
+		return true, nil
+	}
+
+	currentValTyped, err := ConvertIntTypeToUint(currVal)
+	if err != nil {
+		return false, err
+	}
+
+	if newVal <= currentValTyped {
+		return false, nil
+	}
+
+	a.tables[tableName] = val
+	a.written[tableName] = false
+	return true, nil
+}
+
+// Confirm is used when an auto increment has actually been written to a table.
+func (a *autoIncrementTracker) Confirm(tableName string, key interface{}) {
+	a.mu.Lock()
+	defer a.mu.Unlock()
+
+	a.tables[tableName] = key
+	a.written[tableName] = true
```

Comment:
How is this supposed to work? We're setting this to true for all values? What if you have some sessions requesting new keys at the same time that other sessions are writing their old keys?